protocol/exporting-action-tests.md

```diff
@@ -0,0 +1,151 @@
+
+These modifications are likely special cases in the L1 and L2 engine actors,
+to export the produced blocks to e.g. JSON files, and any other data such as state.
+
+Potentially, the op-program DB can be reused in the action-tests,
+to capture witness data, such that individual blocks can be executed in a FP setting.
+
+A Fault-Proof program, when given a L1 chain and L2 chain,
+and potentially witness data, should be able to reproduce the state-transition,
+and verify the results match against the expected L2 post-state.
+
+To do more extensive testing, of e.g. alternative L2 consensus nodes,
+we need to also export the reorg signals and safety signals,
+so these offchain concepts can be acted on by the test target.
+
+For interop, ideally the export format includes test-actor names and chain-IDs,
+so we can support multi-chain tests.
+
+### action-tests DSL
+
+The current "actions" DSL has a few quirks, we may want to improve on:
+- Some actor functions produce an action, which then still has to be executed.
+  The original idea here was to randomize/fuzz actions; this may still be on the table, but will require more time.
+- The test-params / genesis generation is quite limited, due to not being able to customize the allocs.
+    - Different genesis parameters can be applied by running the forge scripts as part of the setup.
+      See `interopgen` work for an example of producing a L1 state and multiple attached L2s, with custom configuration, within a second or two.
+    - Forge scripts could help produce custom pre-states. The test could run these scripts as part of the setup phase, using the same tooling as the above.
+
+These things would be worthwhile to improve as part of the testing effort, benefiting both the Go stack, and quality of exported test vectors.
+
+## Resource Usage
+
+<!-- What is the resource usage of the proposed solution?
+Does it consume a large amount of computational resources or time? -->
+
+The action-tests already run quite rapidly: there is no scheduling to wait for.
+
+However, exporting the test-data will likely add some overhead:
+DB features like witness-generation, and file-IO.
+
+By making this test-data export optional, the regular action-tests can still run quickly in CI.
+
+# Alternatives Considered
+
+<!-- List out a short summary of each possible solution that was considered.
+Comparing the effort of each solution -->
+
+See [testing discussion call notes of 9 Sept 2024](https://www.notion.so/oplabs/Weekly-Sync-09-09-2024-implementation-discussions-for-testing-improvements-103c4ed68cc24137af92aaa115292d06).
+
+# Risks & Uncertainties
+
+<!-- An overview of what could go wrong.
+Also any open questions that need more work to resolve. -->
+
+## Time allocation
+
+The action-tests are a shared responsibility, but primarily maintained by Proto,
+who is occupied with Interop, and not directly involved in Holocene and Proofs work.
+There may be some onboarding required to be effective.
```

Comment:
I think it makes a lot of sense for Holocene Derivation devs (@geoknee, me primarily at this point) and Proofs devs involved in multi-client testing (@clabby @refcell ?) to do the grunt work here, with only guidance provided by Proto. Holocene needs more features from action tests for testing derivation edge cases including out-of-order and invalid frames and batches, and multi-client testing needs the chain data exports.